Don't require an EAD ID from ASpace in order to download resource

marlo-longley opened the pull request:
Closes #666

corylown commented on app/models/aspace/resource.rb:
```diff
@@ -1,23 +1,40 @@
   # Model for an Aspace resource to manage information
   # needed for downloading and updating resources in ArcLight
   class Resource
-    attr_reader :ead_id, :uri, :repository_code
+    attr_reader :ead_id, :uri, :identifier, :repository_code
 
-    def initialize(ead_id:, uri:, repository_code:)
+    def initialize(uri:, repository_code:, ead_id: nil, identifier: nil)
       @ead_id = ead_id
+      @identifier = identifier
       @uri = uri
       @repository_code = repository_code
     end
```
I'd prefer if the parameters/variables were listed in the same order each time so it's easier to understand at a glance that they're the same.

corylown commented on app/models/aspace/resource.rb:
```diff
@@ -1,23 +1,40 @@
+      resource_id.sub(/(\.xml)?$/i, '').parameterize
+    end
+
+    def resource_id
```
`resource_id` could be confused with the aspace resource id. Some other ideas: `call_number`, `file_identifier`, `finding_aid_identifier`? Open to ideas.

corylown replied:
Should this be a private method?

marlo-longley replied:
Good thinking. I like file_identifier I think?

corylown commented on app/models/aspace/resource.rb:
```diff
@@ -1,23 +1,40 @@
+    end
+
+    def resource_id
+      if ead_id
+        ead_id
+      elsif identifier
+        # Identifier can be entered in ASpace by archivists
+        # It often matches the EAD ID, but with variable punctionation or spacing
+        identifier
+      else
+        # If no EAD ID or identifier is provided, throw an error.
+        # We need an identifier to generate a file name and determine the repository code
+        raise AspaceResourceError, 'Resources from ArchivesSpace must have an EAD ID or identifier'
+      end
```
Maybe? With your helpful comments above the method.
```suggestion
            ead_id || identifier || raise(AspaceResourceError,
                                    'Resources from ArchivesSpace must have an EAD ID or identifier')
```

corylown replied:
I think we will want to rescue from this error where it might be encountered in `DownloadEadJob` and log it to honeybadger so that if raised it doesn't interrupt queuing up the rest of the jobs.

marlo-longley commented:
@corylown I addressed your comments, thanks for taking a look. The thing I'm most unsure about is if I am sending stuff to HB in a good way / the error handling.

corylown commented on app/jobs/download_ead_job.rb:
```diff
@@ -81,10 +81,14 @@ def self.enqueue_repository(repository:, config:)
     file_dir = "#{config.data_directory}/#{resource.arclight_repository_code}"
-    DownloadEadJob.perform_later(resource_uri: resource.uri,
-                                 file_name: resource.file_name,
-                                 file_dir:, index: config.index,
-                                 generate_pdf: config.generate_pdf)
+    begin
+      file_name = resource.file_name
+    rescue Aspace::AspaceResourceError => e
+      Rails.logger.debug "Failed to create filename for download: #{e.message}"
+    end
+
+    DownloadEadJob.perform_later(resource_uri: resource.uri, file_name:, file_dir:,
+                                 index: config.index, generate_pdf: config.generate_pdf)
   end
   private_class_method :enqueue_resource
```
I think we want something like:

```
  def self.enqueue_resource(resource:, config:)
    file_dir = "#{config.data_directory}/#{resource.arclight_repository_code}"
    DownloadEadJob.perform_later(resource_uri: resource.uri,
                                 file_name: resource.file_name,
                                 file_dir:, index: config.index,
                                 generate_pdf: config.generate_pdf)
  rescue Aspace::AspaceResourceError => e
    message = "Failed to create filename for #{resource.uri}: #{e.message}"
    Rails.logger.warn(message)
    Honeybadger.notify(message)
  end
  private_class_method :enqueue_resource
```

marlo-longley commented:
@corylown thank you. I added your suggestions for the error handling.

corylown commented on app/services/arclight_repository_mapper.rb:
```diff
@@ -4,27 +4,29 @@
   def map_to_code
     code = if aspace_repository_code == 'speccoll'
-             code_from_ead_id
+             # Fall back on manuscripts
+             # if we can't determine the sub-category of speccoll from the syntax of the file_identifier
+             file_identifier ? code_from_file_identifier : 'manuscripts'
```
We don't need this anymore, right? Since we'll raise `AspaceResourceError` before we get here if we don't have a file identifier?
